# Data Cleaning Pipeline

## Objective
Build a reusable, systematic data cleaning pipeline that transforms raw data into a clean, model-ready dataset.

## Why this matters
- Real-world data is messy
- Models fail more due to bad data than bad algorithms
- Pipelines ensure reproducibility and consistency

In [1]:
import numpy as np
import pandas as pd

from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler, OneHotEncoder
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.impute import SimpleImputer

import warnings
warnings.filterwarnings("ignore")

pd.set_option("display.max_columns", None)


In [2]:
df = pd.read_csv("https://raw.githubusercontent.com/datasciencedojo/datasets/master/titanic.csv")
df.head()



PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   
3            4         1       1   
4            5         0       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   
3       Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                           Allen, Mr. William Henry    male  35.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S  
3      0            113803  53.1000  C123        S  
4      0            373450   8.0500   NaN        S

In [3]:
df.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


In [4]:
df.isna().sum().sort_values(ascending=False)


Cabin          687
Age            177
Embarked         2
PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
SibSp            0
Parch            0
Ticket           0
Fare             0
dtype: int64

# Analysis till Now: 

Missing values :

1. Cabin     -     687
2. Age       -     177
3. Embarked  -       2

a. Which columns are useless - (`Cabin`)

b. Which need imputation - (`Age`, `Embarked`)

c. Which need encoding - (`Sex`, `Embarked`)

In [5]:
# Drop Irrelevant / High-Noise Columns
df = df.drop(columns=["Cabin", "Ticket", "Name", "PassengerId"])


In [6]:
# Separate Features & Target
X = df.drop("Survived", axis=1)
y = df["Survived"]


In [ ]:
# Identify Column Types (Very Important)
numerical_features = X.select_dtypes(include=["int64", "float64"]).columns.tolist()
categorical_features = X.select_dtypes(include=["object"]).columns.tolist()

numerical_features, categorical_features


(['Pclass', 'Age', 'SibSp', 'Parch', 'Fare'], ['Sex', 'Embarked'])

In [11]:
# Build Numerical & Categorical Pipelines
numeric_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="median")),
    ("scaler", StandardScaler())
])


In [9]:
# Categorical Pipeline
categorical_pipeline = Pipeline(steps=[
    ("imputer", SimpleImputer(strategy="most_frequent")),
    ("encoder", OneHotEncoder(handle_unknown="ignore"))
])


In [12]:
# Combine Pipelines (ColumnTransformer)
preprocessor = ColumnTransformer(
    transformers=[
        ("num", numeric_pipeline, numerical_features),
        ("cat", categorical_pipeline, categorical_features)
    ]
)


In [14]:
# Train-Test Split (Before Fitting!)
X_train, X_test, y_train, y_test = train_test_split(
    X, y, test_size=0.2, random_state=42, stratify=y
)


In [15]:
# Apply the Cleaning Pipeline
X_train_clean = preprocessor.fit_transform(X_train)
X_test_clean = preprocessor.transform(X_test)

X_train_clean.shape, X_test_clean.shape


((712, 10), (179, 10))

In [16]:
# Convert to DataFrame (Optional but Nice)
feature_names = (
    numerical_features +
    list(preprocessor.named_transformers_["cat"]
         .named_steps["encoder"]
         .get_feature_names_out(categorical_features))
)

X_train_clean_df = pd.DataFrame(X_train_clean, columns=feature_names)
X_train_clean_df.head()


Pclass       Age     SibSp     Parch      Fare  Sex_female  Sex_male  \
0  0.829568 -0.081135 -0.465084 -0.466183  0.513812         0.0       1.0   
1 -0.370945 -0.081135 -0.465084 -0.466183 -0.662563         0.0       1.0   
2 -1.571457 -0.081135 -0.465084 -0.466183  3.955399         0.0       1.0   
3  0.829568 -0.887827 -0.465084  0.727782 -0.467874         1.0       0.0   
4 -0.370945  0.110934  0.478335  0.727782 -0.115977         1.0       0.0   

   Embarked_C  Embarked_Q  Embarked_S  
0         0.0         0.0         1.0  
1         0.0         0.0         1.0  
2         0.0         0.0         1.0  
3         0.0         0.0         1.0  
4         0.0         0.0         1.0

In [18]:
# Save the Pipeline (Production Mindset)
import joblib

joblib.dump(preprocessor, "data_cleaning_pipeline.pkl")


['data_cleaning_pipeline.pkl']

# Final Summary (Markdown)

## Key Takeaways

- Data cleaning is not ad-hoc; it must be systematic
- Pipelines prevent data leakage
- The same pipeline must be used in training and production
- Good data > complex models
